tools/make_patch.c

```diff
@@ -0,0 +1,532 @@
+#define PROGRAM_NAME "make_patch"
+#define USAGE_OPTS "labels.sym constants.txt patched.gbc original.gbc vc.patch.template vc.patch"
+
+#include "common.h"
+
+#include <ctype.h>
+
+#define HIGH(x) (((x) >> 8) & 0xff)
+#define LOW(x) ((x) & 0xff)
+
+struct Buffer {
+	size_t size;
+	char data[];
+};
+
+struct Patch {
+	unsigned int offset;
+	unsigned int size;
+};
+
+struct Patches {
+	size_t size;
+	size_t capacity;
+	struct Patch *data;
+};
+
+struct Symbol {
+	struct Symbol *next;
+	int address;
+	unsigned int offset;
+	char name[];
+};
+
+struct Buffer *create_buffer(void) {
+	struct Buffer *buffer = xmalloc(sizeof(*buffer) + 0x100);
+	buffer->size = 0x100;
+	return buffer;
+}
+
+struct Buffer *expand_buffer(struct Buffer *buffer, size_t size) {
+	if (size > buffer->size) {
+		size_t new_size = size > buffer->size + 0x100 ? size : buffer->size + 0x100;
+		buffer = xrealloc(buffer, new_size);
+		buffer->size = new_size;
+	}
+	return buffer;
+}
+
+struct Patches *create_patches(void) {
+	struct Patches *patches = xmalloc(sizeof(*patches));
+	patches->size = 0;
+	patches->capacity = 10;
+	patches->data = xmalloc(patches->capacity * sizeof(*patches->data));
+	return patches;
+}
+
+void append_patch(struct Patches *patches, unsigned int offset, unsigned int size) {
+	if (patches->size >= patches->capacity) {
+		patches->capacity = (patches->capacity + 1) * 2;
+		patches->data = xrealloc(patches->data, patches->capacity * sizeof(*patches->data));
+	}
+	patches->data[patches->size++] = (struct Patch){.offset = offset, .size = size};
+}
+
+void free_patches(struct Patches *patches) {
+	free(patches->data);
+	free(patches);
+}
+
+int get_address_type_limit(int address) {
```

Comment:
Consider:

```c
int get_address_type_limit(int address) {
	return address >= 0x8000 && address < 0xa000 ? 0xa000 // VRAM
		: address >= 0xa000 && address < 0xc000 ? 0xc000 // SRAM
		: address >= 0xc000 && address < 0xf000 ? 0xd000 // WRAM
		: address >= 0xff80 && address < 0xffff ? 0x10000 // HRAM
		: 0x4000; // ROM
}
```

Reply:
@mid-kid @vulcandth Is this going too far to save 6 lines?

Reply:
I think it looks cleaner. We can do it.